swap/src/asb/rendezvous.rs

```diff
@@ -0,0 +1,155 @@
+use crate::rendezvous::XmrBtcNamespace;
+use anyhow::Result;
+use libp2p::core::connection::ConnectionId;
+use libp2p::identity::Keypair;
+use libp2p::multiaddr::Protocol;
+use libp2p::rendezvous::{Event, Namespace};
+use libp2p::swarm::{
+    IntoProtocolsHandler, NetworkBehaviour, NetworkBehaviourAction, NetworkBehaviourEventProcess,
+    PollParameters, ProtocolsHandler,
+};
+use libp2p::{Multiaddr, PeerId};
+use std::task::{Context, Poll};
+use std::time::{Duration, Instant};
+
+/// A `NetworkBehaviour` that handles registration of the xmr-btc swap service with a rendezvous point
+pub struct Behaviour {
+    rendezvous_behaviour: libp2p::rendezvous::Rendezvous,
+    rendezvous_point_peer_id: PeerId,
+    rendezvous_point_addr: Multiaddr,
+    rendezvous_namespace: XmrBtcNamespace,
+    rendezvous_reregister_timestamp: Option<Instant>,
+    is_connected: bool,
+    events: Vec<NetworkBehaviourAction<BehaviourInEvent, ()>>,
+}
+
+impl Behaviour {
+    pub fn new(
+        keypair: Keypair,
+        peer_id: PeerId,
+        addr: Multiaddr,
+        namespace: XmrBtcNamespace,
+    ) -> Self {
+        Self {
+            rendezvous_behaviour: libp2p::rendezvous::Rendezvous::new(
+                keypair,
+                libp2p::rendezvous::Config::default(),
+            ),
+            rendezvous_point_peer_id: peer_id,
+            rendezvous_point_addr: addr,
+            rendezvous_namespace: namespace,
+            rendezvous_reregister_timestamp: None,
+            is_connected: false,
+            events: vec![],
+        }
+    }
+
+    pub fn refresh_registration(&mut self) -> Result<()> {
+        if self.is_connected {
+            if let Some(rendezvous_reregister_timestamp) = self.rendezvous_reregister_timestamp {
+                if Instant::now() > rendezvous_reregister_timestamp {
+                    self.rendezvous_behaviour.register(
+                        Namespace::new(self.rendezvous_namespace.to_string())
+                            .expect("our namespace to be a correct string"),
+                        self.rendezvous_point_peer_id,
+                        None,
+                    )?;
+                }
+            }
+        } else {
+            let p2p_suffix = Protocol::P2p(self.rendezvous_point_peer_id.into());
+            let address_with_p2p = if !self
+                .rendezvous_point_addr
+                .ends_with(&Multiaddr::empty().with(p2p_suffix.clone()))
+            {
+                self.rendezvous_point_addr.clone().with(p2p_suffix)
+            } else {
+                self.rendezvous_point_addr.clone()
+            };
+            self.events.push(NetworkBehaviourAction::DialAddress {
+                address: address_with_p2p,
+            })
+        }
+        Ok(())
+    }
+}
+
+type BehaviourInEvent =
+<<<libp2p::rendezvous::Rendezvous as NetworkBehaviour>::ProtocolsHandler as IntoProtocolsHandler>::Handler as ProtocolsHandler>::InEvent;
```

Comment:
Looks like there was a merge conflict here?

Reply:
nope don't think so. It builds